src/signature/commitment_signature.rs

```diff
@@ -0,0 +1,292 @@
+// Copyright 2021 The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+//! Digital Signature module
+//! This module defines generic traits for handling the digital signature operations, agnostic
+//! of the underlying elliptic curve implementation
+
+use crate::{
+    commitment::{HomomorphicCommitment, HomomorphicCommitmentFactory},
+    keys::{PublicKey, SecretKey},
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    cmp::Ordering,
+    hash::{Hash, Hasher},
+    marker::PhantomData,
+    ops::{Add, Mul},
+};
+use tari_utilities::{hex::Hex, ByteArray};
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error, PartialEq, Eq, Deserialize, Serialize)]
+pub enum CommitmentSignatureError {
+    #[error("An invalid challenge was provided")]
+    InvalidChallenge,
+}
+
+#[allow(non_snake_case)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CommitmentSignature<P, K, C> {
+    public_commitment_nonce: HomomorphicCommitment<P>,
+    signature_u: K,
+    signature_v: K,
+    _commitment_factory: PhantomData<C>,
+}
+
+impl<P, K, C> CommitmentSignature<P, K, C>
+where
+    P: PublicKey<K = K>,
+    K: SecretKey,
+    C: HomomorphicCommitmentFactory<P = P> + Default,
```

Comment:
This can probably only be passed to the methods that need it. I suspect that the type of factory does not need to be specified as part of the type.

Reply:
100% This is too opinionated. 

You've gone to the trouble of making the struct very general, but suddenly can't use this without default commitment factories.

Just pass in `&C` in the methods that need a factory, as Mike suggetsed.

Reply:
That is a possibility, but it's  going to make the code messier. 

We use the signatures as: 
```
pub type RistrettoComSig = CommitmentSignature<RistrettoPublicKey, RistrettoSecretKey, PedersenCommitmentFactory>;
RistrettoComSig::sign(a_value_alice, x_value_alice, k_2_alice, k_1_alice, &challenge);
sig.verify_challenge(&commitment, &challenge));
```

If we remove the phantomdata

it's going to look as follows:
```
pub type RistrettoComSig = CommitmentSignature<RistrettoPublicKey, RistrettoSecretKey>;
RistrettoComSig::sign::<PedersenCommitmentFactory>(a_value, x_value, k_2, k_1, &challenge);
sig.verify_challenge::<PedersenCommitmentFactory>(&commitment, &challenge)
```